src/ui/feed/note/notedata.rs

```diff
@@ -0,0 +1,230 @@
+        let mut has_nostr_event_reference = false;
+        for shard in &shattered_content.segments {
+            match shard {
+                ContentSegment::NostrUrl(nurl) => match nurl.0 {
+                    NostrBech32::Id(_) | NostrBech32::EventPointer(_) => {
+                        has_nostr_event_reference = true;
+                    },
+                    _ => (),
+                },
+                ContentSegment::TagReference(_) => { has_tag_reference = true; },
+                ContentSegment::Hyperlink(_) => (),
+                ContentSegment::Plain(_) => (),
+            }
+        }
+
+        let repost = {
+            let content_trim = event.content.trim();
+
+            if event.kind == EventKind::Repost && embedded_event.is_some()
+            {
+                Some(RepostType::Kind6Embedded)
+            } else if has_tag_reference || has_nostr_event_reference || content_trim.is_empty() {
+                if !cached_mentions.is_empty() {
+                    if content_trim.is_empty() {
+                        // handle NIP-18 conform kind:6 with 'e' tag but no content
+                        shattered_content.segments.push(ContentSegment::TagReference(0));
+
+                        if event.kind == EventKind::Repost {
+                            Some(RepostType::Kind6Mention)
+                        } else {
+                            Some(RepostType::MentionOnly)
+                        }
+                    } else {
+                        Some(RepostType::CommentMention)
+                    }
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        };
+
+        // If delegated, use the delegated person
+        let author_pubkey: PublicKeyHex = if let EventDelegation::DelegatedBy(pubkey) = delegation {
+            pubkey.into()
+        } else {
+            event.pubkey.into()
+        };
+
+        let author = match GLOBALS.people.get(&author_pubkey) {
+            Some(p) => p,
+            None => DbPerson::new(author_pubkey),
+        };
+
+        Some(NoteData {
+            event,
+            delegation,
+            author,
+            deletion,
+            repost,
+            embedded_event,
+            cached_mentions,
+            reactions,
+            self_already_reacted,
+            shattered_content,
+        })
+    }
+
+    pub(super) fn update_reactions(&mut self) {
+        let (mut reactions, self_already_reacted) = Globals::get_reactions_sync(self.event.id);
+
+        self.reactions.clear();
+        self.reactions.append(&mut reactions);
+
+        self.self_already_reacted = self_already_reacted;
+    }
+}
+
+/// a 'note' is a processed event
+pub struct Notes {
+    notes: DashMap<Id, NoteData>,
+}
+
+impl Notes {
+    pub fn new() -> Notes {
+        Notes {
+            notes: DashMap::new(),
+        }
+    }
+
+    pub(super) fn get(&self, id: &Id) -> Option<NoteData> {
```

Comment:
@mikedilger I wanted to return a reference to NoteData here, but I couldn't get this to work using lifetime specifiers, I just don't understand them well enough to define that the reference should be valid as long as GossipUi object exists. I did have a solution using `Arc<RwLock<NoteData>>` but it's very clunky and needs way to many calls to the mutex. 
Another solution would be to separate the reactions into an outer wrapper, so that we can use `Arc<NoteData>`, reactions are the only reason to have a mutable NoteData.
What do you think?

Reply:
The UI is single threaded so it should not need locks and it shouldn't Send anything. If you find yourself wanting multiple references to the same thing, you can use RefCell which allows multiple simultaneous immutable borrows, and try_borrow_mut() to mutably borrow it (while nobody else has a borrow).  Ideally you write the code in a way that you never have multiple outstanding borrows at the same time.

These NoteData do not live as long as GossipUi.  GossipUi is created first and destroyed after.  And they get destroyed on every recompute, right?  So that kind of lifetime would never work.  But it doesn't matter, you should just borrow it and return it right away.  If you have to borrow it twice,  then (borrow, return, borrow, return) rather than trying to hold it for longer.

Reply:
That is, borrowing it and passing it down a long chain of functions is unnecessary headache.  Just borrow it multiple times.  The GossipUi code that holds it in a (presumably) RefCell would be very fast.